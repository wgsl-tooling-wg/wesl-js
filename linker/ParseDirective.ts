import {
  anyThrough,
  kind,
  opt,
  or,
  repeat,
  req,
  seq,
  setTraceNames,
  tokens,
  tracing,
} from "@wesl/mini-parse";
import { gleamImport } from "./GleamImport.ts";
import {
  argsTokens,
  moduleTokens,
} from "./MatchWgslD.ts";
import { eolf, makeElem } from "./ParseSupport.ts";

/* parse #directive enhancements to wgsl: #export, etc. */

const argsWord = kind(argsTokens.arg);
const fromWord = or(argsWord, kind(argsTokens.relPath));

const fromClause = seq(
  "from",
  or(fromWord.tag("from"), seq('"', fromWord.tag("from"), '"')),
);

/** #export <foo> <(a,b)> EOL */
export const exportDirective = seq(or("#export", "export"), opt(eolf)).map(
  r => {
    const e = makeElem("export", r, ["args"]);
    r.app.state.push(e);
  },
);

const moduleDirective = seq(
  or("module", "#module"),
  tokens(moduleTokens, req(kind(moduleTokens.moduleName).tag("name"))),
  eolf,
).map(r => {
  const e = makeElem("module", r);
  e.name = normalizeModulePath(r.tags.name[0]);
  r.app.state.push(e);
});

function normalizeModulePath(name: string): string {
  if (name.includes("::")) {
    const result = name.split("::").join("/");
    return result;
  }
  return name;
}

export const directive = tokens(
  argsTokens,
  seq(repeat("\n"), or(exportDirective, gleamImport, moduleDirective)),
);

if (tracing) {
  setTraceNames({
    fromClause,
    exportDirective,
    moduleDirective,
    directive,
  });
}
