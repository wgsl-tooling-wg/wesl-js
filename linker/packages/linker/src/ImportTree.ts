export class ImportTree {
  /** segments in path order */
  constructor(public segments: PathSegment[]) {}
}

export type PathSegment = SimpleSegment | ImportTree | SegmentList;

export class SimpleSegment {
  constructor(
    public name: string,
    public as?: string,
    public args?: string[], // generic args (only allowed on final segment). TODO drop
  ) {}
}

/** or choices for this path segment */
export class SegmentList {
  constructor(public list: PathSegment[]) {}
}

export function treeToString(tree: ImportTree): string {
  return tree.segments.map(s => segmentToString(s)).join("/");
}

function segmentToString(segment: PathSegment): string {
  if (segment instanceof SimpleSegment) {
    const { name, as, args } = segment;
    const asMsg = as ? ` as ${as}` : "";
    const argsMsg = args ? `(${args.join(", ")})` : "";
    return `${name}${argsMsg}${asMsg}`;
  }
  if (segment instanceof SegmentList) {
    return `{${segment.list.map(s => segmentToString(s)).join(", ")}}`;
  }
  if (segment instanceof ImportTree) {
    return `(${treeToString(segment)})`;
  }
  return `|unknown segment type ${(segment as any).constructor.name}|`;
}
