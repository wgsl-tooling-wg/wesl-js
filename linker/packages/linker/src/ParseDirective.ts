import {
  anyThrough,
  kind,
  opt,
  or,
  repeat,
  req,
  seq,
  setTraceNames,
  tokens,
  tracing,
} from "mini-parse";
import { gleamImport } from "./GleamImport.js";
import { argsTokens, lineCommentTokens, mainTokens } from "./MatchWgslD.js";
import { eolf } from "./ParseSupport.js";

/* parse #directive enhancements to wgsl: #export, etc. */

const argsWord = kind(argsTokens.arg);
const fromWord = or(argsWord, kind(argsTokens.relPath));

const fromClause = seq("from", or(fromWord, seq('"', fromWord, '"')));

/** #export <foo> <(a,b)> EOL */
export const exportDirective = seq(or("#export", "export"), opt(eolf));

export const directive = tokens(
  argsTokens,
  seq(repeat("\n"), or(exportDirective, gleamImport)),
);

const skipToEol = tokens(lineCommentTokens, anyThrough(eolf));

/** parse a line comment */
export const lineComment = seq(tokens(mainTokens, "//"), skipToEol);

if (tracing) {
  setTraceNames({
    fromClause,
    exportDirective,
    skipToEol,
    lineComment,
    directive,
  });
}
