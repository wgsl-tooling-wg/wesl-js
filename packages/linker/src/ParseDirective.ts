import {
  anyThrough,
  kind,
  opt,
  or,
  repeat,
  req,
  seq,
  setTraceNames,
  tokens,
  tracing,
} from "mini-parse";
import { gleamImport } from "./GleamImport.js";
import {
  argsTokens,
  lineCommentTokens,
  mainTokens,
  moduleTokens,
} from "./MatchWgslD.js";
import { eolf, makeElem } from "./ParseSupport.js";

/* parse #directive enhancements to wgsl: #export, etc. */

const argsWord = kind(argsTokens.arg);
const fromWord = or(argsWord, kind(argsTokens.relPath));

const fromClause = seq(
  "from",
  or(fromWord.tag("from"), seq('"', fromWord.tag("from"), '"')),
);

/** #export <foo> <(a,b)> EOL */
export const exportDirective = seq(or("#export", "export"), opt(eolf)).map(
  r => {
    const e = makeElem("export", r, ["args"]);
    r.app.state.push(e);
  },
);

const moduleDirective = seq(
  or("module", "#module"),
  tokens(moduleTokens, req(kind(moduleTokens.moduleName).tag("name"))),
  eolf,
).map(r => {
  const e = makeElem("module", r);
  e.name = normalizeModulePath(r.tags.name[0]);
  r.app.state.push(e);
});

function normalizeModulePath(name: string): string {
  if (name.includes("::")) {
    const result = name.split("::").join("/");
    return result;
  }
  return name;
}

export const directive = tokens(
  argsTokens,
  seq(repeat("\n"), or(exportDirective, moduleDirective)),
);

const skipToEol = tokens(lineCommentTokens, anyThrough(eolf));

/** parse a line comment */
export const lineComment = seq(tokens(mainTokens, "//"), skipToEol);

if (tracing) {
  setTraceNames({
    fromClause,
    exportDirective,
    skipToEol,
    lineComment,
    moduleDirective,
    directive,
  });
}
