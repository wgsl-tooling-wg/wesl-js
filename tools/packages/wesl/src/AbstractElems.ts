import { Span } from "mini-parse";
import { DeclIdent, RefIdent, SrcModule } from "./Scope.ts";

/**
 * Structures to describe the 'interesting' parts of a WESL source file.
 *
 * The parts of the source that need to analyze further in the linker
 * are pulled out into these structures.
 *
 * The parts that are uninteresting the the linker are recorded
 * as 'TextElem' nodes, which are generally just copied to the output WGSL
 * along with their containing element.
 */
export type AbstractElem = GrammarElem | SyntheticElem;

export type GrammarElem = ContainerElem | TerminalElem;

export type ContainerElem =
  | AttributeElem
  | AliasElem
  | ConstAssertElem
  | ConstElem
  | UnknownExpressionElem
  | SimpleMemberRef
  | FnElem
  | TypedDeclElem
  | GlobalVarElem
  | LetElem
  | ModuleElem
  | OverrideElem
  | FnParamElem
  | StructElem
  | StructMemberElem
  | StuffElem
  | TypeRefElem
  | VarElem;

/** Inspired by https://github.com/wgsl-tooling-wg/wesl-rs/blob/3b2434eac1b2ebda9eb8bfb25f43d8600d819872/crates/wgsl-parse/src/syntax.rs#L364 */
export type ExpressionElem =
  | Literal
  | TranslateTimeFeature
  | RefIdentElem
  | ParenthesizedExpression
  | ComponentExpression
  | ComponentMemberExpression
  | UnaryExpression
  | BinaryExpression
  | FunctionCallExpression;

export type TerminalElem =
  | DirectiveElem
  | DeclIdentElem //
  | NameElem
  | RefIdentElem
  | TextElem
  | ImportElem;

export type GlobalDeclarationElem =
  | AliasElem
  | ConstElem
  | FnElem
  | GlobalVarElem
  | OverrideElem
  | StructElem;

export type DeclarationElem = GlobalDeclarationElem | FnParamElem | VarElem;

export interface AbstractElemBase {
  kind: AbstractElem["kind"];
  start: number;
  end: number;
}

export interface ElemWithContentsBase extends AbstractElemBase {
  contents: AbstractElem[];
}

/* ------   Terminal Elements  (don't contain other elements)  ------   */

/**
 * a raw bit of text in WESL source that's typically copied to the linked WGSL.
 * e.g. a keyword  like 'var'
 * or a phrase we needn't analyze further like '@diagnostic(off,derivative_uniformity)'
 */
export interface TextElem extends AbstractElemBase {
  kind: "text";
  srcModule: SrcModule;
}

/** a name that doesn't need to be an Ident
 * e.g.
 * - a struct member name
 * - a diagnostic rule name
 * - an enable-extension name
 * - an interpolation sampling name
 */
export interface NameElem extends AbstractElemBase {
  kind: "name";
  name: string;
}

/** an identifier that refers to a declaration (aka a symbol reference) */
export interface RefIdentElem extends AbstractElemBase {
  kind: RefIdent["kind"];
  ident: RefIdent;
  srcModule: SrcModule;
}

/** a declaration identifier (aka a symbol declaration) */
export interface DeclIdentElem extends AbstractElemBase {
  kind: DeclIdent["kind"];
  ident: DeclIdent;
  srcModule: SrcModule;
}

/** Holds an import statement, and has a span */
export interface ImportElem extends AbstractElemBase {
  kind: "import";
  imports: ImportStatement;
}

/**
 * An import statement, which is tree shaped.
 * `import foo::bar::{baz, cat as neko};
 */
export interface ImportStatement {
  kind: "import-statement";
  segments: ImportSegment[];
  finalSegment: ImportCollection | ImportItem;
}

/**
 * A collection of import trees.
 * `{baz, cat as neko}`
 */
export interface ImportSegment {
  kind: "import-segment";
  name: string;
}

/**
 * A primitive segment in an import statement.
 * `foo`
 */
export interface ImportCollection {
  kind: "import-collection";
  subtrees: ImportStatement[];
}

/**
 * A renamed item at the end of an import statement.
 * `cat as neko`
 */
export interface ImportItem {
  kind: "import-item";
  name: string;
  as?: string;
}

/* ------   Synthetic element (for transformations, not produced by grammar) ------   */

/** generated element, produced after parsing and binding */
export interface SyntheticElem {
  kind: "synthetic";
  text: string;
}

/* ------   Container Elements  (contain other elements)  ------   */

/** a declaration identifer with a possible type */
export interface TypedDeclElem extends ElemWithContentsBase {
  kind: "typeDecl";
  decl: DeclIdentElem;
  typeRef?: TypeRefElem; // TODO Consider a variant for fn params and alias where typeRef is required
}

/** an alias statement */
export interface AliasElem extends ElemWithContentsBase {
  kind: "alias";
  name: DeclIdentElem;
  typeRef: TypeRefElem;
}

/** an attribute like '@compute' or '@binding(0)' */
export interface AttributeElem extends ElemWithContentsBase {
  kind: "attribute";
  attribute: Attribute;
}
export type Attribute =
  | StandardAttribute
  | InterpolateAttribute
  | BuiltinAttribute
  | DiagnosticAttribute
  | IfAttribute;
export interface StandardAttribute {
  kind: "attribute";
  name: string;
  params: UnknownExpressionElem[];
}
export interface InterpolateAttribute {
  kind: "@interpolate";
  params: NameElem[];
}
export interface BuiltinAttribute {
  kind: "@builtin";
  param: NameElem;
}
export interface DiagnosticAttribute {
  kind: "@diagnostic";
  severity: NameElem;
  rule: [NameElem, NameElem | null];
}
export interface IfAttribute {
  kind: "@if";
  param: TranslateTimeExpressionElem;
}

/** a const_assert statement */
export interface ConstAssertElem extends ElemWithContentsBase {
  kind: "assert";
}

/** a const declaration */
export interface ConstElem extends ElemWithContentsBase {
  kind: "const";
  name: TypedDeclElem;
}

export interface UnknownExpressionElem extends ElemWithContentsBase {
  kind: "expression";
}

export interface TranslateTimeExpressionElem {
  kind: "translate-time-expression";
  expression: ExpressionElem;
  span: Span;
}

/** A literal value in WESL source. A boolean or a number. */
export interface Literal {
  kind: "literal";
  value: string;
  span: Span;
}

/** `words`s inside `@if` */
export interface TranslateTimeFeature {
  kind: "translate-time-feature";
  name: string;
  span: Span;
}

/** (expr) */
export interface ParenthesizedExpression {
  kind: "parenthesized-expression";
  expression: ExpressionElem;
}
/** `foo[expr]` */
export interface ComponentExpression {
  kind: "component-expression";
  base: ExpressionElem;
  access: ExpressionElem;
}
/** `foo.member` */
export interface ComponentMemberExpression {
  kind: "component-member-expression";
  base: ExpressionElem;
  access: NameElem;
}
/** `+foo` */
export interface UnaryExpression {
  kind: "unary-expression";
  operator: UnaryOperator;
  expression: ExpressionElem;
}
/** `foo + bar` */
export interface BinaryExpression {
  kind: "binary-expression";
  operator: BinaryOperator;
  left: ExpressionElem;
  right: ExpressionElem;
}
/** `foo(arg, arg)` */
export interface FunctionCallExpression {
  kind: "call-expression";
  function: RefIdentElem;
  arguments: ExpressionElem[];
}
export interface UnaryOperator {
  value: "!" | "&" | "*" | "-" | "~";
  span: Span;
}
export interface BinaryOperator {
  value:
    | ("||" | "&&" | "+" | "-" | "*" | "/" | "%" | "==")
    | ("!=" | "<" | "<=" | ">" | ">=" | "|" | "&" | "^")
    | ("<<" | ">>");
  span: Span;
}

export interface DirectiveElem extends AbstractElemBase {
  kind: "directive";
  directive: DiagnosticDirective | EnableDirective | RequiresDirective;
}

export interface DiagnosticDirective {
  kind: "diagnostic";
  severity: NameElem;
  rule: [NameElem, NameElem | null];
}
export interface EnableDirective {
  kind: "enable";
  extensions: NameElem[];
}
export interface RequiresDirective {
  kind: "requires";
  extensions: NameElem[];
}

/** a function declaration */
export interface FnElem extends ElemWithContentsBase {
  kind: "fn";
  name: DeclIdentElem;
  params: FnParamElem[];
  fnAttributes?: AttributeElem[];
  returnType?: TypeRefElem;
}

/** a global variable declaration (at the root level) */
export interface GlobalVarElem extends ElemWithContentsBase {
  kind: "gvar";
  name: TypedDeclElem;
}

/** an entire file */
export interface ModuleElem extends ElemWithContentsBase {
  kind: "module";
}

/** an override declaration */
export interface OverrideElem extends ElemWithContentsBase {
  kind: "override";
  name: TypedDeclElem;
}

/** a parameter in a function declaration */
export interface FnParamElem extends ElemWithContentsBase {
  kind: "param";
  name: TypedDeclElem;
}

/** simple references to structures, like myStruct.bar
 * (used for transforming refs to binding structs) */
export interface SimpleMemberRef extends ElemWithContentsBase {
  kind: "memberRef";
  name: RefIdentElem;
  member: NameElem;
  extraComponents?: StuffElem;
}

/** a struct declaration */
export interface StructElem extends ElemWithContentsBase {
  kind: "struct";
  name: DeclIdentElem;
  members: StructMemberElem[];
  bindingStruct?: true; // used later during binding struct transformation
}

/** generic container of other elements */
export interface StuffElem extends ElemWithContentsBase {
  kind: "stuff";
}

/** a struct declaration that's been marked as a bindingStruct */
export interface BindingStructElem extends StructElem {
  bindingStruct: true;
  entryFn?: FnElem;
}

/** a member of a struct declaration */
export interface StructMemberElem extends ElemWithContentsBase {
  kind: "member";
  name: NameElem;
  attributes?: AttributeElem[];
  typeRef: TypeRefElem;
  mangledVarName?: string; // root name if transformed to a var (for binding struct transformation)
}

export type TypeTemplateParameter = TypeRefElem | UnknownExpressionElem;

/** a reference to a type, like 'f32', or 'MyStruct', or 'ptr<storage, array<f32>, read_only>'   */
export interface TypeRefElem extends ElemWithContentsBase {
  kind: "type";
  name: RefIdent;
  templateParams?: TypeTemplateParameter[];
}

/** a variable declaration */
export interface VarElem extends ElemWithContentsBase {
  kind: "var";
  name: TypedDeclElem;
}

export interface LetElem extends ElemWithContentsBase {
  kind: "let";
  name: TypedDeclElem;
}
