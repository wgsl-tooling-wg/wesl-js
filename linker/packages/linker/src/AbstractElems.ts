import { ImportTree } from "./ImportTree.ts";
import { DeclIdent, RefIdent, SrcModule } from "./Scope.ts";

/** 
 * Structures to describe the 'interesting' parts of a WESL source file.
 *
 * The parts of the source that need to analyze further in the linker 
 * are pulled out into these structures.
 * 
 * The parts that are uninteresting the the linker are recorded
 * as 'TextElem' nodes, which are generally just copied to the output WGSL
 * along with their containing element.
 */
export type AbstractElem =
  | AliasElem
  | AttributeElem
  | ConstElem
  | ExpressionElem
  | ImportElem
  | ConstAssertElem
  | FnElem
  | RefIdentElem
  | DeclIdentElem
  | ModuleElem
  | NameElem
  | OverrideElem
  | ParamElem
  | StructElem
  | StructMemberElem
  | TextElem
  | TypeRefElem
  | GlobalVarElem
  | VarElem;

export type DeclarationElem =
  | AliasElem
  | ConstElem
  | OverrideElem
  | ParamElem
  | FnElem
  | StructElem
  | GlobalVarElem
  | VarElem;

export interface AbstractElemBase {
  kind: string;
  start: number;
  end: number;
}

export interface ElemWithContents extends AbstractElemBase {
  contents: AbstractElem[];
}

/** a raw bit of text in WESL source that's typically copied to the linked WGSL. 
 e.g. a keyword  like 'var' 
 or a phrase we needn't analyze further like '@diagnostic(off,derivative_uniformity)'
*/
export interface TextElem extends AbstractElemBase {
  kind: "text";
  srcModule: SrcModule;
}

export interface ImportElem extends ElemWithContents {
  kind: "import";
  imports: ImportTree;
}

export type TypeTemplateParameter = TypeRefElem | ExpressionElem | string;

/** an expression (generally we don't need details of expressions, just their contained idents */
export interface ExpressionElem extends ElemWithContents {
  kind: "expression";
}

/** a reference to a type, like 'f32', or 'MyStruct', or 'ptr<storage, array<f32>, read_only>'   */
export interface TypeRefElem extends ElemWithContents {
  kind: "type";
  name: RefIdent | string;
  templateParams?: TypeTemplateParameter[];
}

/** an identifier that refers to a declaration */
export interface RefIdentElem extends AbstractElemBase {
  kind: RefIdent["kind"];
  ident: RefIdent;
  srcModule: SrcModule;
}

/** a declaration identifier */
export interface DeclIdentElem extends AbstractElemBase {
  kind: DeclIdent["kind"];
  ident: DeclIdent;
  srcModule: SrcModule;
}

/** a parameter in a function declaration */
export interface ParamElem extends ElemWithContents {
  kind: "param";
  name: DeclIdentElem;
  typeRef: TypeRefElem;
}

/** an attribute like '@compute' or '@binding(0)' */
export interface AttributeElem extends ElemWithContents {
  kind: "attribute";
  name: string;
  params?: ExpressionElem[];
}

/** a variable declaration */
export interface VarElem extends ElemWithContents {
  kind: "var";
  name: DeclIdentElem;
  typeRef?: TypeRefElem;
}

/** a global variable declaration (at the root level) */
export interface GlobalVarElem extends ElemWithContents {
  kind: "gvar";
  name: DeclIdentElem;
  typeRef?: TypeRefElem;
}

/** a const declaration */
export interface ConstElem extends ElemWithContents {
  kind: "const";
  name: DeclIdentElem;
  typeRef?: TypeRefElem;
}

/** an override declaration */
export interface OverrideElem extends ElemWithContents {
  kind: "override";
  name: DeclIdentElem;
  typeRef?: TypeRefElem;
}

/** an entire file */
export interface ModuleElem extends ElemWithContents {
  kind: "module";
}

/** an alias statement */
export interface AliasElem extends ElemWithContents {
  kind: "alias";
  name: DeclIdentElem;
  typeRef: TypeRefElem;
}

/** a const_assert statement */
export interface ConstAssertElem extends ElemWithContents {
  kind: "assert";
}

/** a struct declaration */
export interface StructElem extends ElemWithContents {
  kind: "struct";
  name: DeclIdentElem;
  members: StructMemberElem[];
  bindingStruct?: true; // used later during binding struct transformation
}

/** a member of a struct declaration */
export interface StructMemberElem extends ElemWithContents {
  kind: "member";
  name: NameElem;
  attributes?: AttributeElem[];  
  typeRef: TypeRefElem;
}

/** a name (e.g. a struct member name) that doesn't need to be an Ident */
export interface NameElem extends AbstractElemBase {
  kind: "name";
  name: string;
  srcModule: SrcModule;
}

/** a function declaration */
export interface FnElem extends ElemWithContents {
  kind: "fn";
  name: DeclIdentElem;
  params: ParamElem[];
  returnType?: TypeRefElem;
}
