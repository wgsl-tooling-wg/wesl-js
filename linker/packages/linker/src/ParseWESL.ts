import { AppState, matchingLexer, ParserInit, SrcMap } from "mini-parse";
import { AbstractElem } from "./AbstractElems.ts";
import { AbstractElem2, ModuleElem } from "./AbstractElems2.ts";
import { mainTokens } from "./MatchWgslD.ts";
import { emptyScope, resetScopeIds, Scope, SrcModule } from "./Scope.ts";
import { OpenElem } from "./WESLCollect.ts";
import { weslRoot } from "./WESLGrammar.ts";

/** result of a parse */
export interface WeslAST {
  elems: AbstractElem[]; // legacy
  rootModule: ModuleElem;
  scope: Scope;
}

/** stable and unstable state used during parsing */
export interface WeslParseState extends AppState<WeslParseContext> {
  context: WeslParseContext;
  stable: StableState;
}

/** stable values used or accumulated during parsing */
export interface StableState {
  // parameters for evaluating conditions while parsing this module
  conditions: Record<string, any>;

  // legacy elems succesfully parsed in this module
  elems: AbstractElem[];

  // elems succesfully parsed in this module
  rootModule?: ModuleElem;

  // root scope for this module
  rootScope: Scope;
}

/** unstable values used during parse collection */
export interface WeslParseContext {
  scope: Scope; // current scope (points somewhere in rootScope)
  openElems: OpenElem[]; // elems that are collecting their contents
}

export function parseSrcModule(
  srcModule: SrcModule,
  maxParseCount: number | undefined = undefined,
): WeslAST {
  const lexer = matchingLexer(srcModule.src, mainTokens);

  const appState = blankWeslParseState();

  const init: ParserInit = { lexer, appState, maxParseCount };
  const parseResult = weslRoot.parse(init);
  if (parseResult === null) {
    throw new Error("parseWESL failed");
  }

  const { rootModule, elems, rootScope } = appState.stable;
  return { rootModule: rootModule!, scope: rootScope, elems };
}

// TODO make wrapper on srcModule
export function parseWESL(
  src: string,
  srcMap?: SrcMap,
  conditions: Record<string, any> = {},
  maxParseCount: number | undefined = undefined,
  grammar = weslRoot,
): WeslAST {
  // TODO allow returning undefined for failure, or throw?

  resetScopeIds();
  const lexer = matchingLexer(src, mainTokens);

  const appState = blankWeslParseState();

  const init: ParserInit = { lexer, appState: appState, srcMap, maxParseCount };
  const parseResult = grammar.parse(init);
  if (parseResult === null) {
    throw new Error("parseWESL failed");
  }

  const { rootModule, elems, rootScope } = appState.stable;
  return { rootModule: rootModule!, scope: rootScope, elems };
}

export function blankWeslParseState(): WeslParseState {
  const rootScope = emptyScope("module");
  return {
    context: { scope: rootScope, openElems: [] },
    stable: { conditions: {}, elems: [], rootScope },
  };
}
