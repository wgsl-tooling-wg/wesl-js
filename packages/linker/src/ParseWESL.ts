import { matchingLexer, ParserInit, SrcMap } from "mini-parse";
import { AbstractElem } from "./AbstractElems.ts";
import { mainTokens } from "./MatchWgslD.ts";
import { WeslParseContext, weslRoot } from "./WESLGrammar.ts";

export function parseWESL(
  src: string,
  srcMap?: SrcMap,
  params: Record<string, any> = {},
  maxParseCount: number | undefined = undefined,
  grammar = weslRoot,
): AbstractElem[] {
  const lexer = matchingLexer(src, mainTokens);
  // state
  const state: AbstractElem[] = [];

  // context is reset on parse failure during backtracking
  const context: WeslParseContext = {
    params,
    provisionalIdents: [],
    provisionalScopes: [],
  };
  const app = {
    context,
    state,
  };
  const init: ParserInit = {
    lexer,
    app,
    srcMap,
    maxParseCount,
  };

  grammar.parse(init);

  return app.state;
}
