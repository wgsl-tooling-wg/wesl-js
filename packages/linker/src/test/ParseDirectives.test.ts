import { expect, test } from "vitest";
import { ModuleElem, TreeImportElem } from "../AbstractElems.js";
import { treeToString } from "../ImportTree.js";
import { directive } from "../ParseDirective.js";
import { parseWESL } from "../ParseWESL.js";
import { testAppParse } from "./TestUtil.js";

test("directive parses #export", () => {
  const { appState } = testAppParse(directive, "#export");
  expect(appState[0].kind).toBe("export");
});

test("parse #export", () => {
  const parsed = parseWESL("#export");
  expect(parsed[0].kind).toBe("export");
});

test("parse import foo/bar", () => {
  const parsed = parseWESL("import foo/bar");
  expect(parsed).toMatchSnapshot();
});

test("parse module foo.bar.ca", () => {
  const src = `module foo.bar.ca`;
  const appState = parseWESL(src);
  expect(appState[0].kind).toBe("module");
  expect((appState[0] as ModuleElem).name).toBe("foo.bar.ca");
});

test("module foo.bar.ca", ctx => {
  const appState = parseWESL(ctx.task.name);
  expect(appState[0].kind).toBe("module");
  expect((appState[0] as ModuleElem).name).toBe("foo.bar.ca");
});

test("module foo::bar::ba", ctx => {
  const appState = parseWESL(ctx.task.name);
  expect(appState[0].kind).toBe("module");
  expect((appState[0] as ModuleElem).name).toBe("foo/bar/ba");
});

test("module foo/bar/ba", ctx => {
  const appState = parseWESL(ctx.task.name);
  expect(appState[0].kind).toBe("module");
  expect((appState[0] as ModuleElem).name).toBe("foo/bar/ba");
});

test("import ./util/foo;", ctx => {
  const appState = parseWESL(ctx.task.name);
  const importElem = appState[0] as TreeImportElem;
  const segments = treeToString(importElem.imports);
  expect(segments).toBe("./util/foo");
});

test("import ./bar/foo;", ctx => {
  const appState = parseWESL(ctx.task.name);
  const importElem = appState[0] as TreeImportElem;
  const segments = treeToString(importElem.imports);
  expect(segments).toBe("./bar/foo");
});

test("import ./bar/{foo,bar};", ctx => {
  const appState = parseWESL(ctx.task.name);
  const imports = appState.filter(e => e.kind === "treeImport");
  const segments = imports.map(i => treeToString(i.imports));
  expect(segments).toContain("./bar/{(foo), (bar)}");
});
