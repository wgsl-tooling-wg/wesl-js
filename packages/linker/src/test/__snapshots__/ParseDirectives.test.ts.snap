// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse import foo/bar 1`] = `
[
  {
    "end": 14,
    "imports": ImportTree {
      "segments": [
        SimpleSegment {
          "args": undefined,
          "as": undefined,
          "name": "foo",
        },
        SimpleSegment {
          "args": undefined,
          "as": undefined,
          "name": "bar",
        },
      ],
    },
    "kind": "treeImport",
    "start": 0,
  },
]
`;
